Add triangle_math tests and fix Triangle3d::bounding_sphere bug

aristaeus opened the pull request:
# Objective

Adopted #12659.

Resolved the merge conflicts on #12659;

* I merged the `triangle_tests` added by this PR and by #13020.
* I moved the [commented out code](https://github.com/bevyengine/bevy/pull/12659#discussion_r1536640427) from the original PR into a separate test with `#[should_panic]`.

alice-i-cecile commented:
Diff on the final commit looks good to me :)

IQuick143 reviewed (changes requested):
I believe that the computation for the degenerate triangle should be done in the same way as for the obtuse one, which will remove a division by zero bug as well as making the code more regular. (Also depending on what optimisations the compiler misses it might make this path faster by avoiding unnecessary math.)

mockersf commented on crates/bevy_math/src/bounding/bounded3d/primitive_impls.rs:
```diff
@@ -607,4 +610,74 @@ mod tests {
+        );
+    }
+
+    #[test]
+    #[should_panic]
+    fn degenerate_bounding_sphere_should_panic() {
+        let zero_degenerate_triangle = Triangle3d::new(Vec3::ZERO, Vec3::ZERO, Vec3::ZERO);
+        let bs = zero_degenerate_triangle.bounding_sphere(Vec3::ZERO, Quat::IDENTITY); // DIVISION BY ZERO
+        assert_eq!(
+            bs.center,
+            Vec3::ZERO.into(),
+            "incorrect bounding sphere center"
+        );
+        assert_eq!(bs.sphere.radius, 0.0, "incorrect bounding sphere radius");
+    }
```
`should_panic` tests shouldn't have asserts

aristaeus replied:
I've changed the test since fixing the panic

aristaeus commented on crates/bevy_math/src/primitives/dim3.rs:
```diff
@@ -1210,8 +1210,86 @@ mod tests {
     }
 
+    #[test]
+    fn extrusion_math() {
+        let circle = Circle::new(0.75);
+        let cylinder = Extrusion::new(circle, 2.5);
+        assert_eq!(cylinder.area(), 15.315264, "incorrect surface area");
+        assert_eq!(cylinder.volume(), 4.417865, "incorrect volume");
+
+        let annulus = crate::primitives::Annulus::new(0.25, 1.375);
+        let tube = Extrusion::new(annulus, 0.333);
+        assert_eq!(tube.area(), 14.886437, "incorrect surface area");
+        assert_eq!(tube.volume(), 1.9124937, "incorrect volume");
+
+        let polygon = crate::primitives::RegularPolygon::new(3.8, 7);
+        let regular_prism = Extrusion::new(polygon, 1.25);
+        assert_eq!(regular_prism.area(), 107.8808, "incorrect surface area");
+        assert_eq!(regular_prism.volume(), 49.392204, "incorrect volume");
```
They're from the original PR. Should I check them?

aristaeus commented on crates/bevy_math/src/bounding/bounded3d/primitive_impls.rs:
```diff
@@ -326,8 +326,9 @@ impl Bounded3d for Triangle3d {
         if self.is_degenerate() {
             let (p1, p2) = self.largest_side();
+            let mid_point = (p1 + p2) / 2.0;
             let (segment, _) = Segment3d::from_points(p1, p2);
```
Ah, sorry, I didn't understand your comments when I read the original PR but I do now. Fixed.

aristaeus commented:
I've simplified the code paths as mentioned, and also added `is_obtuse()` (and `is_acute` for completeness).

I've just realised that `Triangle2d` doesn't have `is_obtuse`, `is_acute` (added in this PR) or `is_degenerate` (already exists). Should I copy those methods over for consistency?

IQuick143 commented on crates/bevy_math/src/primitives/dim3.rs:
```diff
@@ -780,6 +780,34 @@ impl Triangle3d {
+        let ca = a - c;
+
+        // a^2 + b^2 < c^2 for an acute triangle
+        let mut side_lengths = [ab.length(), bc.length(), ca.length()];
```
You can avoid unecessary square roots by querying for `length_squared` instead of `length` (since you square the result anyway, and the comparison operations are unaffected by squaring since squaring positive numbers is a rising function)

IQuick143 commented on crates/bevy_math/src/primitives/dim3.rs:
```diff
@@ -780,6 +780,34 @@ impl Triangle3d {
+    /// Checks if the triangle is obtuse, meaning one angle is greater than 90 degrees
+    #[inline(always)]
+    pub fn is_obtuse(&self) -> bool {
+        let [a, b, c] = self.vertices;
+        let ab = b - a;
+        let bc = c - b;
+        let ca = a - c;
+
+        // a^2 + b^2 > c^2 for an obtuse triangle
+        let mut side_lengths = [ab.length(), bc.length(), ca.length()];
+        side_lengths.sort_by(|a, b| a.partial_cmp(b).unwrap());
+        side_lengths[0].powf(2.) + side_lengths[1].powf(2.) < side_lengths[2].powf(2.)
```
An alternative implementation would be to take the dot products between sides and check if the sign is negative at any angle.

I'm not exactly sure that's necessarily a better implementation, so I think this is good, provided we remove the unnecessary square roots hidden in the `.length()` calls.

IQuick143 commented on crates/bevy_math/src/bounding/bounded3d/primitive_impls.rs:
```diff
@@ -323,29 +323,15 @@ impl Bounded3d for Triangle3d {
     /// that contains the largest side of the triangle.
-    fn bounding_sphere(&self, translation: Vec3, rotation: Quat) -> BoundingSphere {
-        if self.is_degenerate() {
+    fn bounding_sphere(&self, translation: Vec3, _: Quat) -> BoundingSphere {
```
Not an expert on this, but maybe the unused parameter should be 
```suggestion
    fn bounding_sphere(&self, translation: Vec3, _rotation: Quat) -> BoundingSphere {
```
so it's better documented.

IQuick143 commented:
Thanks for the work on this PR!
